internal/controller/nonadminbackupstoragelocation_controller.go

```diff
@@ -163,6 +164,42 @@ func (r *NonAdminBackupStorageLocationReconciler) initNaBSLDelete(ctx context.Co
 	return false, nil
 }
 
+// deleteNonAdminBackups deletes all NonAdminBackups associated with the given NonAdminBackupStorageLocation
+func (r *NonAdminBackupStorageLocationReconciler) deleteNonAdminBackups(ctx context.Context, logger logr.Logger, nabsl *nacv1alpha1.NonAdminBackupStorageLocation) (bool, error) {
+	nonAdminBackupList := &nacv1alpha1.NonAdminBackupList{}
+	listOpts := &client.ListOptions{Namespace: nabsl.Namespace}
+
+	if err := r.Client.List(ctx, nonAdminBackupList, listOpts); err != nil {
+		return false, err
+	}
+
+	if len(nonAdminBackupList.Items) == 0 {
+		logger.V(1).Info("No NonAdminBackups found in NonAdminBackupStorageLocation namespace", "nabsl", nabsl.Name)
+		return false, nil
+	}
+
+	for _, nonAdminBackup := range nonAdminBackupList.Items {
+		// Ensure it belongs to this StorageLocation
+		if nonAdminBackup.Spec.BackupSpec == nil || nonAdminBackup.Spec.BackupSpec.StorageLocation != nabsl.Name {
```

Comment:
Ideally this check should be nabsl.uuid

Reply:
Otherwise this code will act on failed to delete NABs from following
> [When the delete() call on the Non Admin Backup object from the NaBSL controller succeded, however during the deletion of Velero Backup triggered by the delete() call fails for some reason, then the NaBSL object will get removed, but the Non Admin Backup object will not get finalizer removal and will not be deleted](https://github.com/migtools/oadp-non-admin/pull/163/files#r1937852423)

Reply:
NABs should have annotations of its NaBSL's UUID which you can reference. Don't add UUID to NAB spec.

Reply:
And you can use NaBSL kube object UUID for this, shouldn't need to add a new uuid field to nabsl spec either.

Reply:
NaBSL is the object created by the user and not the NaBSL controller, so I am not sure why we would want to generate UUID for it? It will be unique name as the user can not have two same objects within the same namespace.

The UUIDs are important to make relation in the oadp namespace between created by the controller Velero Backup and Velero BSL objects.

Reply:
We store the NaBSL UUID as part of the NAB Status, so user can not change it, we could use it, but as above I am not sure if it will give us anything?

https://github.com/migtools/oadp-non-admin/blob/11abd720930b2899b709027eed86773d424c4372/internal/controller/nonadminbackupstoragelocation_controller.go#L315-L320

Similar way of storing UUIDs is for Non Admin Backup and Non Admin Restore

Reply:
> And you can use NaBSL kube object UUID for this, shouldn't need to add a new uuid field to nabsl spec either.

You are (probably) right. This is mostly to be similar across all NAC objects. For Velero Backup we actually need other then UUID from cluster, because when Velero Sync Controller recreates that Velero Backup object (e.g. on a new cluster) the UUID is different and we can not map it back to NaBSL objects which created them.

Reply:
I guess NaBSL.status.uuid is already a thing..

Reply:
> why we would want to generate UUID for it?

I never said that.

Reply:
> I guess NaBSL.status.uuid is already a thing..

We generate this on a first NaBSL creation and then we reference it in the Velero BSL. Do you think it's wrong?

Reply:
> It will be unique name as the user can not have two same objects within the same namespace.

User can delete NaBSL and recreate. Without checking for matching NaBSL uuid on the NAB annotations, you can potentially be deleting NAB on NaBSLCreatedInMarch when NAC should only delete NAB for NaBSLCreatedInApril.

Reply:
> We generate this on a first NaBSL creation and then we reference it in the Velero BSL. Do you think it's wrong?
For UUID meant for avoiding collisions in the same cluster, I believe kube provided is sufficient and object.status.uuid is redundant.

If we have cross cluster logic to avoid collision of NaBSL then yes NaBSL.status.UUID could proved useful. Then there needs to be mechanics to upload UUID to the bucket IMO to "save" the UUID for the bucket.

Reply:
I think this will not happen. The user deletes NaBSL based on name, so user is always deleting one object. The controller is removing all the dependencies based on generated UUIDs, which for both user objects are different as we don't generate UUID based on name/namespace, but we actually make it unique and guarantee they will never be same (collision probability is extremely minimal - UUID4 with additional prefix)

Reply:
> The controller is removing all the dependencies based on generated UUIDs

only if NAC is running.

Reply:
Many kube users overcommit their clusters, there is always a chance a controller is not running.

Reply:
Ah right so you are suggesting to include UUID information from NaBSL inside the NAB object and also ensure they do match. Got it! In normal operation this should not happen as the NaBSL removal should remove Backups when NAC is running, but I see your point ! Think this would be follow up work ?

Reply:
Yes.